src/libraries/System.Console/src/System/IO/StdInReader.cs

```diff
@@ -363,5 +373,26 @@ private unsafe ConsoleKeyInfo ReadKey()
 
         /// <summary>Gets whether there's input waiting on stdin.</summary>
         internal static bool StdinReady => Interop.Sys.StdinReady();
+
+        private void EchoToTerminal(char c)
+        {
+            Span<byte> bytes = stackalloc byte[32]; // 32 bytes seems ample
+            int bytesWritten = 1;
+            if (Ascii.IsValid(c))
+            {
+                bytes[0] = (byte)c;
+            }
+            else
+            {
+                var chars = new ReadOnlySpan<char>(in c);
+                bytesWritten = _echoEncoder.GetBytes(chars, bytes, flush: false);
+                if (bytesWritten == 0)
+                {
+                    return;
+                }
+            }
```

Comment:
@adamsitnik does this look good to you? I wrote it this way to have the `bytes` `Span` is in the same scope as the `WriteToTerminal` call.